Add self-contained artifact upload script for apache nexus (#11329)

dweiss opened the pull request.

dweiss committed:
Wrote a no-dependency Java script that uploads to Apache Nexus. Modified the release wizard. Heavily inspired by Houston's work on Solr side.

dweiss committed:
Flatten.

dweiss commented:
I initially wanted to use Python but without additional libraries it overwhelmed me and I wanted to keep it simple and self-contained. There is some extra verbosity (XML processing in Java is hell) but it does work and I staged artifacts for Lucene 10.0.0 successfully (and dropped the repository afterwards).

Now... I'm not so sure how to test the wizard bit. I based the changes on @HoustonPutman 's work on Solr side. The script can accept user name and password in environment variables, on command line or will prompt for password if not provided... but I haven't tested this as I'm not sure how to test it in isolation (@janhoy - is there any way to do it?).

dweiss commented:
For stand-alone use, it's simple enough too:
```
java dev-tools\scripts\StageArtifacts.java -u dweiss /release/candidate/maven-artifacts
```
will prompt for password for nexus access.

jpountz commented:
Thanks @dweiss I'll give it a try!

dweiss committed:
Add ASL headers.

madrob commented on dev-tools/scripts/StageArtifacts.java:
```diff
@@ -0,0 +1,395 @@
+              break;
+            case "-u":
+            case "--user":
+              params.userName = args[++i];
```
Can we create a separate method for getting the next arg that would throw something nicer than an IOOBException when somebody (i.e. me) inevitably goofs up?

dweiss replied:
there's a catch block that does it, below. it's a waste of time to elaborate much on arg validation if it's used from the wizard anyway?

dweiss replied:
I added that method, just for you, @madrob :)

madrob commented on dev-tools/scripts/StageArtifacts.java:
```diff
@@ -0,0 +1,395 @@
+
+            case "-h":
+            case "--help":
+              System.out.println("java " + StageArtifacts.class.getName() + " [options] path");
+              System.out.println("  -u, --user  User name for authentication.");
+              System.out.println("              better: ASF_USERNAME env. var.");
+              System.out.println("  -p, --password  Password for authentication.");
+              System.out.println("              better: ASF_PASSWORD env. var.");
+              System.out.println("  -n, --nexus URL to Apache Nexus (optional).");
+              System.out.println("  --description  Staging repo description (optional).");
+              System.out.println("");
+              System.out.println("  path        Path to maven artifact directory.");
+              System.out.println("");
+              System.out.println(" Password can be omitted for console prompt-input.");
```
Use multi-line string?

madrob commented on dev-tools/scripts/StageArtifacts.java:
```diff
@@ -0,0 +1,395 @@
+            // we could use json if XML is too difficult to work with.
+            // .header("Accept", "application/json")
+            .build();
+        HttpResponse<String> response = client.send(req, bodyHandler);
```
do we need retries?

dweiss replied:
I don't think so. if something fails, it'll fail badly - nexus is not graceful here...

madrob commented on dev-tools/scripts/StageArtifacts.java:
```diff
@@ -0,0 +1,395 @@
+            .filter(Files::isRegularFile)
+            // Ignore locally generated maven metadata files.
+            .filter(path -> !path.getFileName().toString().startsWith("maven-metadata."))
+            .sorted(Comparator.comparing(Path::toString))
```
why? do we need parent projects first or something like that?

dweiss committed:
Add a more explicit error in case an option is left without an argument.

dweiss commented:
@jpountz What did you point the script at, exactly? I'll look into it - drop the staging repo, please.

dweiss commented:
I think I know what the problem is, but would like a confirmation from you first - let me know what you pointed the script at (which directory).

dweiss commented:
The script should be pointed at the maven folder, including parent subfolders, as in:

https://dist.apache.org/repos/dist/dev/lucene/lucene-9.4.0-RC2-rev-0384b4fcad7856ddc574c8b994c814a568ce6d0a/lucene/maven/

I'm surprised nexus even closed the staging repository since artifact group IDs shouldn't allow you to do that...

jpountz commented:
This is the folder I ran it against. I used the same folder that the release wizard recommended using with the Ant build.

dweiss commented:
Give me a sec. you're getting a different output than I do then.

dweiss commented:
Hi @jpountz . Sorry about this - I'm not sure what happened there but the relative upload path of those artifacts was indeed wrong (and I'm also not sure why nexus isn't complaining about it!...). I fixed the script on 9x and main and added an extra check to make sure the groupId matches every uploaded artifact's path.

You can also add {{--description "Lucene 9x"}} to the command line - it shows as the staging repository's description in the UI.

jpountz commented:
No worries at all @dweiss, thanks a lot for modernizing Maven staging upload, it's a relief we no longer have a dependency on a 8.x build!

dweiss commented:
I wonder if this should be reported to infra though. The fact you can close an upload repository with top-level artifacts is strange... I wonder what would happen if you clicked release on that, would it actually promote those JARs to here? It seems crazy!
https://repo1.maven.org/maven2/

jpountz commented:
@dweiss So I did release the staged repository with top-level artifacts, and Nexus did not give me any errors either. :( After browsing the content, I saw all the artifacts that I had expected, it's only after clicking and realizing that publication to Central wasn't working that I thought that the lack of hierarchy was weird.

Good news is that I have released with the correct directory layout earlier today and now I can see the artifacts on Maven Central.

dweiss commented:
I filed this question to infra: https://issues.apache.org/jira/browse/INFRA-23931

jpountz commented:
Thanks Dawid!

dweiss commented:
Just FYI - Sonatype's nexus wouldn't let you release that staging repository, I've tried:
![image](https://user-images.githubusercontent.com/199470/203274558-06abae96-ee76-4b56-b919-a0a73aa0d6c5.png)

dweiss commented:
I'll follow up with Gavin on the Infra ticket. Seems like the rules at Apache are not as strict. And sorry for this.

jpountz commented:
Again, no worries at all, not your fault.
